compiler/rustc_lint/src/if_let_rescope.rs

```diff
@@ -0,0 +1,305 @@
+    /// a significant destructor, this lint remains only a suggestion and is set to `allow` by default.
+    ///
+    /// Whenever possible, a rewrite into an equivalent `match` expression that
+    /// observe the same order of calls to such destructors is proposed by this lint.
+    /// Authors may take their own discretion whether the rewrite suggestion shall be
+    /// accepted, or rejected to continue the use of the `if let` expression.
+    pub IF_LET_RESCOPE,
+    Allow,
+    "`if let` assigns a shorter lifetime to temporary values being pattern-matched against in Edition 2024 and \
+    rewriting in `match` is an option to preserve the semantics up to Edition 2021",
+    @future_incompatible = FutureIncompatibleInfo {
+        reason: FutureIncompatibilityReason::EditionSemanticsChange(Edition::Edition2024),
+        reference: "issue #124085 <https://github.com/rust-lang/rust/issues/124085>",
+    };
+}
+
+declare_lint_pass!(
+    /// Lint for potential change in program semantics of `if let`s
+    IfLetRescope => [IF_LET_RESCOPE]
+);
+
+impl<'tcx> LateLintPass<'tcx> for IfLetRescope {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>) {
+        if !expr.span.edition().at_least_rust_2021() || !cx.tcx.features().if_let_rescope {
+            return;
+        }
+        let hir::ExprKind::If(cond, conseq, alt) = expr.kind else { return };
+        let hir::ExprKind::Let(&hir::LetExpr {
+            span,
+            pat,
+            init,
+            ty: ty_ascription,
+            recovered: Recovered::No,
+        }) = cond.kind
+        else {
+            return;
+        };
+        let source_map = cx.tcx.sess.source_map();
+        let expr_end = expr.span.shrink_to_hi();
+        let if_let_pat = expr.span.shrink_to_lo().between(init.span);
+        let before_conseq = conseq.span.shrink_to_lo();
+        let lifetime_end = source_map.end_point(conseq.span);
+
+        if let ControlFlow::Break(significant_dropper) =
+            (FindSignificantDropper { cx }).visit_expr(init)
+        {
+            let lint_without_suggestion = || {
+                cx.tcx.emit_node_span_lint(
+                    IF_LET_RESCOPE,
+                    expr.hir_id,
+                    span,
+                    IfLetRescopeRewrite { significant_dropper, lifetime_end, sugg: None },
+                )
+            };
+            if ty_ascription.is_some() || !expr.span.can_be_used_for_suggestions() {
+                // Our `match` rewrites does not support type ascription,
+                // so we just bail.
+                // Alternatively when the span comes from proc macro expansion,
+                // we will also bail.
+                // FIXME(#101728): change this when type ascription syntax is stabilized again
+                lint_without_suggestion();
+            } else {
+                let Ok(pat) = source_map.span_to_snippet(pat.span) else {
```

Comment:
Question: do we need to gate this `pat.span` with `can_be_used_for_suggestions` too? Since I want to say we gated the outer `expr.span`, but I'm not sure if we did for the `pat.span`. Note that `can_be_used_for_suggestions` is not aware of AST/HIR tree structures, so just because `expr.span` can be used for suggestion does not necessarily mean `pat.span` can be too AFAIK, unfortunately.

https://github.com/rust-lang/rust/blob/f167efad2f51088d86180ee89177b3d7c9e7c2f5/compiler/rustc_span/src/lib.rs#L579-L587

Reply:
Yeah, let's take caution and gate them with `pat.span` as well. 👍